readthedocs/config/config.py

```diff
@@ -745,12 +751,60 @@ def validate_build(self):
                 ),
             )
 
-        # Allow to override specific project
-        config_image = self.defaults.get('build_image')
-        if config_image:
-            build['image'] = config_image
+            # Allow to override specific project
+            config_image = self.defaults.get('build_image')
+            if config_image:
+                build['image'] = config_image
+
+        with self.catch_validation_error('build.apt_packages'):
+            raw_packages = self._raw_config.get('build', {}).get('apt_packages', [])
+            validate_list(raw_packages)
+            # Transform to a dict, so is easy to validate individual entries.
+            self._raw_config.setdefault('build', {})['apt_packages'] = (
+                list_to_dict(raw_packages)
+            )
+
+            build['apt_packages'] = [
+                self.validate_apt_package(index)
+                for index in range(len(raw_packages))
+            ]
+            if not raw_packages:
+                self.pop_config('build.apt_packages')
+
         return build
 
+    def validate_apt_package(self, index):
```

Comment:
I feel like we probably can be smarter about this. Can we not call apt after a `--`, which is the shell way of handling the end of options? 

> Unless otherwise noted, each builtin command documented as accepting options preceded by ‘-’ accepts ‘--’ to signify the end of the options. The :, true, false, and test/[ builtins do not accept options and do not treat ‘--’ specially. The exit, logout, return, break, continue, let, and shift builtins accept and process arguments beginning with ‘-’ without requiring ‘--’. Other builtins that accept arguments but are not specified as accepting options interpret arguments beginning with ‘-’ as invalid options and require ‘--’ to prevent this interpretation.

We should still raise a validation error here, but I'd like to see additional restrictions for this built in.

Reply:
Good point Eric.

Also, we need to take care of executing commands in the same line as well. Like,

```
apt-get install `touch /tmp/hello.txt`
```

Reply:
I believe python's subprocess command will handle this, but we should definitely be sure and have tests for it.

Reply:
Yes, `subprocess` should handle this as long as you don't pass `shell=True`. However, it's worth a test to make sure we don't ever mess that up.

Reply:
I don't think we use subprocess, but the Docker client API (`exec_create`), tho.

Reply:
We don't use subprocess, we escape special chars in https://github.com/readthedocs/readthedocs.org/blob/2435a4a09837ef0e91e2fa1cc6869498311525cc/readthedocs/doc_builder/environments.py#L335-L367

Reply:
Have you tested defining packages in a malicious way? (e.g. trying to executed forbidden things). From this docstring it only happen under certain circumstances. Are we sure we are always calling this in that way?

Reply:
`escape_command` defaults to true if that's what you mean.